From 48c657f23a61a41a46842b25bce4f287a56223a2 Mon Sep 17 00:00:00 2001
From: Takayuki 'January June' Suwa <jjsuwa_sys3175@yahoo.co.jp>
Date: Mon, 20 Jun 2022 01:56:16 +0900
Subject: [PATCH 30/31] xtensa: Fix RTL insn cost estimation about relaxed MOVI
 instructions

These instructions will all be converted to L32R ones with litpool entries
by the assembler.

gcc/ChangeLog:

	* config/xtensa/xtensa.c (xtensa_is_insn_L32R_p):
	Consider relaxed MOVI instructions as L32R.
---
 gcc/config/xtensa/xtensa.c | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/gcc/config/xtensa/xtensa.c b/gcc/config/xtensa/xtensa.c
index 595c5f96f..b92ec9caa 100644
--- a/gcc/config/xtensa/xtensa.c
+++ b/gcc/config/xtensa/xtensa.c
@@ -4322,17 +4322,23 @@ xtensa_is_insn_L32R_p (const rtx_insn *insn)
 {
   rtx x = PATTERN (insn);
 
-  if (GET_CODE (x) == SET)
+  if (GET_CODE (x) != SET)
+    return false;
+
+  x = XEXP (x, 1);
+  if (MEM_P (x))
     {
-      x = SET_SRC (x);
-      if (MEM_P (x))
-	{
-	  x = XEXP (x, 0);
-	  return (SYMBOL_REF_P (x) || CONST_INT_P (x))
-		 && CONSTANT_POOL_ADDRESS_P (x);
-	}
+      x = XEXP (x, 0);
+      return (SYMBOL_REF_P (x) || CONST_INT_P (x))
+	     && CONSTANT_POOL_ADDRESS_P (x);
     }
 
+  /* relaxed MOVI instructions, that will be converted to L32R by the
+     assembler.  */
+  if (CONST_INT_P (x)
+      && ! xtensa_simm12b (INTVAL (x)))
+    return true;
+
   return false;
 }
 
-- 
2.20.1

